web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts

```diff
@@ -352,13 +352,18 @@ export class DocumentsService {
    */
   update(
     uri: DocumentUri,
-    updated: Partial<Document> | ((document: Document) => Document)
+    updated: Partial<Document> | ((document: Document) => void | Document)
   ) {
     this.setState(draft => {
       const index = draft.documents.findIndex(doc => doc.uri === uri);
       const toUpdate = draft.documents[index];
       if (typeof updated === 'function') {
-        draft.documents[index] = updated(toUpdate);
+        const newState = updated(toUpdate);
+        // if a new object was returned,
+        // we have to set it at the appropriate index
+        if (newState) {
+          draft.documents[index] = newState;
+        }
       }
       Object.assign(toUpdate, updated);
```

Comment:
I originally asked "Shouldn't we call `Object.assign` only if `updated` is not a function?", but…

…now that I think about it, isn't it a bit weird that calling `update` with an object does not follow Immer API, but calling `update` with a function does? In general, I'd avoid having a method that has different behavior depending on the type of the argument being passed.

I think we should either have a separate method which accepts a function (and follows Immer API) or stick with `updated: Partial<Document> | (document: Document) => void`. I don't think there's a place in the project where we return a completely new state in Immer. Narrowing the function to the mutable one ensures that the `update` method conceptually works the same way, no matter if you pass an object or a function.

Reply:
Makes sense, done.